### ISSUE TITLE
Research what requires so much memory on the ingester during persist

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#4886]
We saw in https://github.com/influxdata/conductor/issues/970 that something (we think it is the ingester persist operation) on a partition that had less than `300MB` of write data consumed over 30GB of RAM and caused our ingesters to crash.

This ticket tracks investigating the root cause of what was consuming so much memory

**Comment by alamb** [comment]
My high level plan for this is to add tracing to the plans run by the ingester (so I can see what the plans are doing)

I am trying to understand what is taking so much memory in the ingester (see https://github.com/influxdata/conductor/issues/970). I suspect it is the datafusion plans, and I want to confirm that. To do so I am hope to use the existing tracing mechanism

We already have a pretty sophisticated tracing mechanism to see what the plans are doing (connected to the datafusion statistics), but they are not hooked up to the ingester (either for user queries or for automated background tasks like `persist` and `compact` operations.

Possible subtasks: 
1. Add ingester information to querier tracing (to see where time is spent)
3. A way to create spans for internal operations like persist
2. Add a way to generate spans for the ingester's persist operations with some random frequency

**Comment by alamb** [comment]
See it happening again https://github.com/influxdata/conductor/issues/980

**Comment by alamb** [comment]
I  spent some time trying to reproduce the large memory explosion we were seeing in the ingester on prod but was not able to.

# What I tried

What I did was use the code on the https://github.com/influxdata/influxdb_iox/tree/alamb/memory_profile branch, which effectively turns off compaction with something like

```diff
diff --git a/influxdb_iox/src/commands/run/all_in_one.rs b/influxdb_iox/src/commands/run/all_in_one.rs
index 9fa3e623b..8b00407fa 100644
--- a/influxdb_iox/src/commands/run/all_in_one.rs
+++ b/influxdb_iox/src/commands/run/all_in_one.rs
@@ -502,17 +502,17 @@ pub async fn command(config: Config) -> Result<()> {
     )
     .await?;

-    info!("starting compactor");
-    let compactor = create_compactor_server_type(
-        &common_state,
-        Arc::clone(&metrics),
-        Arc::clone(&catalog),
-        Arc::clone(&object_store),
-        Arc::clone(&exec),
-        Arc::clone(&time_provider),
-        compactor_config,
-    )
-    .await?;
+    // info!("starting compactor");
+    // let compactor = create_compactor_server_type(
+    //     &common_state,
+    //     Arc::clone(&metrics),
+    //     Arc::clone(&catalog),
+    //     Arc::clone(&object_store),
+    //     Arc::clone(&exec),
+    //     Arc::clone(&time_provider),
+    //     compactor_config,
+    // )
+    // .await?;

     let ingester_addresses = vec![format!("http://{}", ingester_run_config.grpc_bind_address)];
     info!(?ingester_addresses, "starting querier");
@@ -533,7 +533,7 @@ pub async fn command(config: Config) -> Result<()> {
     let services = vec![
         Service::create(router, &router_run_config),
         Service::create_grpc_only(ingester, &ingester_run_config),
-        Service::create_grpc_only(compactor, &compactor_run_config),
+        //Service::create_grpc_only(compactor, &compactor_run_config),
         Service::create_grpc_only(querier, &querier_run_config),
     ];
```

Then I ran `influxdb_iox` like this:

```shell
TMPDIR=/data/tmp LOG_FILTER=ingester=debug,info cargo run --release --no-default-features --features=heappy  -- run all-in-one --max-http-request-size=1000000000
```

And tried in a loop to stuff it full of data, hoping to provoke the large memory expansion we saw on produduction. However, I was not able to

Next steps:
1. I will try and extract some actual data from one of the tables that was causing the issue and replay that through my local ingester setup to see if I can reproduce the issue

**Comment by pauldix** [comment]
Ah, adding my idea for refinement on the persist knobs here in public:
> I think that an additional per partition persist threshold would help with this particular case. One that is a limit on the number of records in a partition before persisting. Add INFLUXDB_IOX_PERSIST_PARTITION_RECORD_THRESHOLD_COUNT and set it to maybe 5M. Then persistence will trigger either when a partition gets too big in estimated size, or when it gets too many records.

> This latter would limit the size explosion on persist for data that is particularly well compressed in the MB format. Also, we generally don't want too many rows buffered up in any individual partition in the ingester.

**Comment by tustvold** [comment]
Perhaps we need to prioritize working on https://github.com/apache/arrow-rs/issues/1764

**Comment by alamb** [comment]
> Ah, adding my idea for refinement on the persist knobs here in public:

FYI @pauldix  -- I agree this knob would be a good way to good temporary stopgap if we can't find an expedient way to cut down on memory usage

**Comment by alamb** [comment]
# Update

TLDR; I think we need at least a factor of 2 between the ingester pause limit and the memory allocated to an IOx ingester. https://github.com/influxdata/influxdb_iox/issues/4936 probably mislead me into thinking there was a larger memory spike than there was. 

## Conclusions
I have spent the day beating on IOx on a gcp machine. Some conclusions
1. I can reproduce similar memory usage behavior we saw on the production ingesters using synthetic load 🎉 
2. I can not reproduce any crazy memory usage explosion (e.g. a factor of 10) with reasonable data
3. The overall `ingester pause` setting seems to be doing an effective job at containing OOMs, even with a single high throughput partition and low cardinality tags. Assuming memory overhead for `persist` is limited (I measured about 2x overhead for reasonable tables) this is good
2. While rapidly ingesting data, individual partitions can end up with much higher buffered amounts of data than the per partition limit. This is not ideal, but I couldn't find evidence of it triggering massive memory use that would trigger OOMs --  thus I am not confident that https://github.com/influxdata/influxdb_iox/pull/4931 would stop many OOMs 
3. With truly outrageous input shapes (100s of tags, tag values that are 1000s of characters log, with very low cardinality) I can probably cause OOMs regardless of the protection settings


## Methodology:
 I have a hacky tool (https://github.com/alamb/low_card): that lets me  send arbitrary amounts of pathalogical, low cardinality line protocol to IOx

The data looks like the following (and can can make more tags, or more measurements)

```
m tag0=A_FAIRLY_LONG_TAG_VALUE,....,tag9=A_FAIRLY_LONG_TAG_VALUE field=4 122
```

Note that  NUM_TAGS=10, and a value length of ~20 is comparable to actual customer data I looked at previously.

With various parameter fiddling, I can causes substantial, but not outrageous memory blowup during persist. I found persist jos that easily consumes 2x the memory of ). I found that a single high throughput partition is the worst.

I ran  IOx on a gcp machine as follows  (with a total persist threshold of 3G and a pause threshold of 4G) and even when hammering it with load could only get the RSS size to peak around 8G.

```shell
TMPDIR=/data/tmp LOG_FILTER=ingester=debug,info cargo run --release  -- run all-in-one  --data-dir=/data/ingester_loading/iox-data-dir --catalog-dsn=postgresql://alamb:alamb@localhost:5432/postgres --pause-ingest-size-bytes=4000000000 \
--persist-memory-threshold-bytes=3000000000 --persist-partition-size-threshold-bytes=300000000 --max-http-request-size=1000000000                                                                                                           
```

Using this branch / rev:
origin/alamb/disable_compactor_all_in_one
638b866cd60216c6aef976c39e9e25eaeb0374b7


## Notes for myself:
I can reset IOx state like thisL

```shell
rm -rf /data/ingester_loading/iox-data-dir/
psql postgres -c 'drop schema iox_catalog cascade';
```

Ran the tool like

```shell
nice cargo run --release
```

**Comment by tustvold** [comment]
I can reproduce @alamb 's findings, and we can see why if we look at what the ArrowWriter is doing - https://github.com/apache/arrow-rs/blob/master/parquet/src/arrow/arrow_writer/mod.rs#L184. Specifically it buffers up all the RecordBatch for a given row group, up to 1M rows, but it writes each `RecordBatch` separately. As a result the memory explosion is bounded by the size of the RecordBatch, which in IOx is 1000 (which seems rather low) as opposed to the row group size.

**Comment by alamb** [comment]
I think we have basically wrapped up the research phase of this ticket. The answer is there is an approximately 2x memory overhead to persist data. This is driven largely by the `Sort` during the persist operation

While we can improve the parquet writer and sort efficiently, we will likely have to improve the memory efficiency of the DataFusion sort if we want to make major improvements.
